File: output/GPT4o-Mini/xCodeEval_remove_4000/C++/Go/f0414e1a73cd1d414131e1a533817dc9.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
	"sort"
)

type DSum struct {
	N, K   int
	A      [][]int64
	A_sum  []int64
	best   int64
}

func (solver *DSum) addItem(knapsack []int64, size int, sum int64) {
	for k := solver.K - size; k >= 0; k-- {
		knapsack[k+size] = max(knapsack[k+size], knapsack[k]+sum)
	}
}

func (solver *DSum) recurse(start, end int, knapsack []int64) {
	if start >= end {
		return
	}

	if end-start == 1 {
		sum := int64(0)
		solver.best = max(solver.best, knapsack[solver.K])

		for prefix := 1; prefix <= len(solver.A[start]); prefix++ {
			sum += solver.A[start][prefix-1]
			solver.best = max(solver.best, sum+knapsack[solver.K-prefix])
		}

		return
	}

	mid := (start + end) / 2
	state := make([]int64, len(knapsack))
	copy(state, knapsack)

	for i := start; i < mid; i++ {
		solver.addItem(state, len(solver.A[i]), solver.A_sum[i])
	}

	solver.recurse(mid, end, state)
	state = make([]int64, len(knapsack))
	copy(state, knapsack)

	for i := mid; i < end; i++ {
		solver.addItem(state, len(solver.A[i]), solver.A_sum[i])
	}

	solver.recurse(start, mid, state)
}

func (solver *DSum) solve() {
	reader := bufio.NewReader(os.Stdin)
	fmt.Fscan(reader, &solver.N, &solver.K)
	solver.A = make([][]int64, solver.N)
	solver.A_sum = make([]int64, solver.N)

	for i := 0; i < solver.N; i++ {
		var t int
		fmt.Fscan(reader, &t)
		solver.A[i] = make([]int64, t)

		for j := 0; j < t; j++ {
			fmt.Fscan(reader, &solver.A[i][j])
		}

		if t > solver.K {
			solver.A[i] = solver.A[i][:solver.K]
		}

		solver.A_sum[i] = sum(solver.A[i])
	}

	solver.best = 0
	knapsack := make([]int64, solver.K+1)
	solver.recurse(0, solver.N, knapsack)
	fmt.Println(solver.best)
}

func sum(arr []int64) int64 {
	total := int64(0)
	for _, v := range arr {
		total += v
	}
	return total
}

func max(a, b int64) int64 {
	if a > b {
		return a
	}
	return b
}

func main() {
	solver := DSum{}
	solver.solve()
}

File: output/GPT4o-Mini/xCodeEval_4000/C++/Java/f0414e1a73cd1d414131e1a533817dc9.java
import java.util.*;
import java.io.*;

class DSum {
    static class Solver {
        private final BufferedReader in;
        private final PrintWriter out;
        private int N, K;
        private List<List<Long>> A;
        private long[] A_sum;
        private long best;

        Solver(BufferedReader in, PrintWriter out) {
            this.in = in;
            this.out = out;
        }

        private void rd() throws IOException {
            StringTokenizer st = new StringTokenizer(in.readLine());
            N = Integer.parseInt(st.nextToken());
            K = Integer.parseInt(st.nextToken());
        }

        private void prln(long result) {
            out.println(result);
        }

        private void addItem(long[] knapsack, int size, long sum) {
            for (int k = K - size; k >= 0; k--) {
                knapsack[k + size] = Math.max(knapsack[k + size], knapsack[k] + sum);
            }
        }

        private void recurse(int start, int end, long[] knapsack) {
            if (start >= end) return;

            if (end - start == 1) {
                long sum = 0;
                best = Math.max(best, knapsack[K]);

                for (int prefix = 1; prefix <= A.get(start).size(); prefix++) {
                    sum += A.get(start).get(prefix - 1);
                    best = Math.max(best, sum + knapsack[K - prefix]);
                }
                return;
            }

            int mid = (start + end) / 2;
            long[] state = knapsack.clone();

            for (int i = start; i < mid; i++) {
                addItem(state, A.get(i).size(), A_sum[i]);
            }

            recurse(mid, end, state);
            state = knapsack.clone();

            for (int i = mid; i < end; i++) {
                addItem(state, A.get(i).size(), A_sum[i]);
            }

            recurse(start, mid, state);
        }

        private void solve() throws IOException {
            rd();
            A = new ArrayList<>(N);
            A_sum = new long[N];

            for (int i = 0; i < N; i++) {
                StringTokenizer st = new StringTokenizer(in.readLine());
                int t = Integer.parseInt(st.nextToken());
                List<Long> list = new ArrayList<>();

                for (int j = 0; j < t; j++) {
                    list.add(Long.parseLong(st.nextToken()));
                }

                if (t > K) {
                    list = list.subList(0, K);
                }

                A.add(list);
                A_sum[i] = list.stream().mapToLong(Long::longValue).sum();
            }

            best = 0;
            long[] knapsack = new long[K + 1];
            recurse(0, N, knapsack);
            prln(best);
        }
    }

    public static void solve(InputStream inputStream, OutputStream outputStream) throws IOException {
        BufferedReader in = new BufferedReader(new InputStreamReader(inputStream));
        PrintWriter out = new PrintWriter(outputStream);
        new Solver(in, out).solve();
        out.flush();
    }

    public static void main(String[] args) throws IOException {
        solve(System.in, System.out);
    }
}
